### ISSUE TITLE
Additional parameters for IIFE output

### ISSUE THREAD

**Issue by ygoe** [rollup/rollup#1962]
How can I specify custom parameters for the IIFE output? rollup will always generate a function with no parameters and call it with no parameters. I need something like this:

    (function ($, window, document, undefined) {
    // code here
    }(jQuery, window, document));

And yes, the argument number mismatch is intentional. undefined shall be undefined, it doesn't need to be passed.

I've tried tracing it in the source code but can't follow it far enough (not even out of the same file, iife.ts).

The --globals and --external arguments don't change a thing and I doubt they wil be able to produce what I'm looking for. Maybe I need to add some tricky `import` statements to my code but I don't know how to write them.

(I guess I'm supposed to ask this in the chat but everybody's away there, so I'm asking where I can better track it for a longer time.)

Edit (15 Feb): I've tried with something like `import $ from 'jQuery';` and adding the argument `--globals "$:jQuery"` but it doesn't work correctly. I can choose from having the immediate function passed `jQuery` (correct) or `$` (wrong) but the function parameter is always named `jQuery` (useless) and never `$` (expected) plus there's additional code generated that doesn't belong there. So right now the bundling isn't usable as-is. I might try patching the generated bundle file, hoping it won't break the source map.

**Comment by jessethomson** [comment]
I am also looking to modify the iife signature. However, in my case, I'd like to remove one of the iife inputs.

**Comment by TrySound** [comment]
Does this work for you guys?
https://rollupjs.org/repl?version=1.2.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMHdpbmRvdyUyMGZyb20lMjAnd2luZG93JyUzQiU1Q25pbXBvcnQlMjBkb2N1bWVudCUyMGZyb20lMjAnZG9jdW1lbnQnJTNCJTVDbmltcG9ydCUyMCUyNCUyMGZyb20lMjAnanF1ZXJ5JyUzQiU1Q24lNUNuY29uc29sZS5sb2cod2luZG93JTJDJTIwZG9jdW1lbnQlMkMlMjAlMjQpJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyaWlmZSUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiUyMmpxdWVyeSUyMiUzQSUyMmpRdWVyeSUyMiUyQyUyMndpbmRvdyUyMiUzQSUyMndpbmRvdyUyMiUyQyUyMmRvY3VtZW50JTIyJTNBJTIyZG9jdW1lbnQlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

**Comment by bugs181** [comment]
Came here to ask this exact question. I have a custom iife expression to expose a `global` object cross-platform.

It looks like this:
```
(function (globalThis, factory) {
  if (typeof exports === 'object' && typeof module !== 'undefined')
    if (typeof global === 'object')
      return factory(global)
    else
      return factory(exports)

  if (typeof define === 'function' && define.amd)
    return define(['exports'], factory)

  factory(globalThis.window = globalThis.window || {})
}(this, function(exports) {
  // user code should go here
}))
```

Currently, the temporary fix for me is to use the `banner` and `footer` properties in the config. However this creates indentation problems that breaks my linter.

I've tried to write the iife above as my `main.js` file however `this` becomes `undefined`, along with no wrapping of the user code.

**Comment by shellscape** [comment]
Hey folks. This is a saved-form message, but rest assured we mean every word. The Rollup team is attempting to clean up the Issues backlog in the hopes that the active and still-needed, still-relevant issues bubble up to the surface. With that, we're closing issues that have been open for an eon or two, and have gone stale like pirate hard-tack without activity.

We really appreciate the folks have taken the time to open and comment on this issue. Please don't confuse this closure with us not caring or dismissing your issue, feature request, discussion, or report. The issue will still be here, just in a closed state. If the issue pertains to a bug, please re-test for the bug on the latest version of Rollup and if present, please tag @shellscape and request a re-open, and we'll be happy to oblige.

**Comment by cpietrzykowski** [comment]
@shellscape I think it makes sense to reopen this. The solution from @TrySound does produce the desired output, but we end up with imports to non-existent modules. If there's another way to specify iife parameters (from globals), I have not found the documents.

**Comment by lukastaegert** [comment]
The solution suggested by @TrySound is the intended way of doing it. The variables passed to the wrapper function correspond to the external imports you have. You should not forget to add `external: ['window', 'document', 'jquery']` to your Rollup config to prevent warnings when bundling.

Adding more configurability here feels for me to be a likely recipe for confusion.

**Comment by cpietrzykowski** [comment]
Yeah, spent a bit more time thinking on this. Not going to fight convention, I don't know if this is a possibility in any other bundler. I got too focused on the output -- expecting everything to wrapped in a bow so to speak. Importing a fake 'document' seems silly, and still feels wrong, however, especially in the case of browser/platform globals, totally unnecessary.

I definitely agree that increasing configuration complexity is the wrong play.

**Comment by lukastaegert** [comment]
I must admit I do not really understand why you would want to import `window` and `document` anyway. As they are globals, you can just use them in your code without they being part of the wrapper? Rollup will automatically take care of deconflicting local variables with the same names if necessary.

**Comment by ygoe** [comment]
Globals can't be minified, locals can.

**Comment by rifler** [comment]
i am sorry for bumping this thread, but is there a solution?

**Comment by cforsyth68** [comment]
> Does this work for you guys?
> https://rollupjs.org/repl?version=1.2.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMHdpbmRvdyUyMGZyb20lMjAnd2luZG93JyUzQiU1Q25pbXBvcnQlMjBkb2N1bWVudCUyMGZyb20lMjAnZG9jdW1lbnQnJTNCJTVDbmltcG9ydCUyMCUyNCUyMGZyb20lMjAnanF1ZXJ5JyUzQiU1Q24lNUNuY29uc29sZS5sb2cod2luZG93JTJDJTIwZG9jdW1lbnQlMkMlMjAlMjQpJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyaWlmZSUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiUyMmpxdWVyeSUyMiUzQSUyMmpRdWVyeSUyMiUyQyUyMndpbmRvdyUyMiUzQSUyMndpbmRvdyUyMiUyQyUyMmRvY3VtZW50JTIyJTNBJTIyZG9jdW1lbnQlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

Not working for me, can you please provide new example?

**Comment by StoneCypher** [comment]
@lukastaegert - please consider disabling the bot that keeps closing all these active tickets

**Comment by alvarotrigo** [comment]
So, any way to add the variable into the IIFE so it can get minified within our code too?

**Comment by alvarotrigo** [comment]
@ygoe  did you find the way?

**Comment by ygoe** [comment]
No. [My tool does post-processing](https://github.com/ygoe/MiniWebCompiler/blob/06363f99843c/MiniWebCompiler/ViewModels/ProjectFile.cs#L258) of the bundled code with scoped regexes before minification to achieve this. I'm not aware* that rollup could do this on its own.

*) Based on my knowledge of over a year ago that I haven't updated since then.

**Comment by alvarotrigo** [comment]
> The solution suggested by @TrySound is the intended way of doing it. The variables passed to the wrapper function correspond to the external imports you have. You should not forget to add `external: ['window', 'document', 'jquery']` to your Rollup config to prevent warnings when bundling.
> 
> Adding more configurability here feels for me to be a likely recipe for confusion.

In my case, the whole purpose of including variables like `window`, `document` or `Math` on the params is just to allow those variables to get minimised. 

The current out put by rollup looks like this:

```js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('window'), require('document'), require('jquery')) :
	typeof define === 'function' && define.amd ? define(['window', 'document', 'jquery'], factory) :
	(global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.window, global.document, global.jQuery));
})(this, (function (window, document, $) { 'use strict';

	function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

	var window__default = /*#__PURE__*/_interopDefaultLegacy(window);
	var document__default = /*#__PURE__*/_interopDefaultLegacy(document);
	var $__default = /*#__PURE__*/_interopDefaultLegacy($);

	console.log(window__default["default"], document__default["default"], $__default["default"]);

}));
``` 

[Link to Rollup reproduction](https://rollupjs.org/repl/?version=2.70.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMHdpbmRvdyUyMGZyb20lMjAnd2luZG93JyUzQiU1Q25pbXBvcnQlMjBkb2N1bWVudCUyMGZyb20lMjAnZG9jdW1lbnQnJTNCJTVDbmltcG9ydCUyMCUyNCUyMGZyb20lMjAnanF1ZXJ5JyUzQiU1Q24lNUNuY29uc29sZS5sb2cod2luZG93JTJDJTIwZG9jdW1lbnQlMkMlMjAlMjQpJTVDbndpbmRvdy5hbGVydCgnZGVtbycpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIybW9kdWxlXzEuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMnVtZCUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiUyMmpxdWVyeSUyMiUzQSUyMmpRdWVyeSUyMiUyQyUyMndpbmRvdyUyMiUzQSUyMndpbmRvdyUyMiUyQyUyMmRvY3VtZW50JTIyJTNBJTIyZG9jdW1lbnQlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==)

So you see, now, instead of using for example:

```js

var a = window;
a.alert('demo');
``` 

Rollup creates:

```js
function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var window__default = /*#__PURE__*/_interopDefaultLegacy(window);

window__default["default"].alert('demo');
``` 

Which ends up making the output result even bigger. As those properties won't get minified. 

Basically, useless for my use case :)
